lib/vrl/stdlib/src/redact.rs

```diff
@@ -39,170 +58,185 @@ impl Function for Redact {
+        match value {
+            Value::Object(object) => {
+                let r#type = match object
+                    .get("type")
+                    .ok_or("filters specified as objects must have type paramater")?
+                {
+                    Value::Bytes(bytes) => Ok(bytes.clone()),
+                    _ => Err("type key in filters must be a string"),
+                }?;
+
+                match r#type.as_ref() {
+                    b"us_social_security_number" => Ok(Filter::UsSocialSecurityNumber),
+                    b"pattern" => {
+                        let patterns = match object
+                            .get("patterns")
+                            .ok_or("pattern filter must have `patterns` specified")?
+                        {
+                            Value::Array(array) => Ok(array
+                                .iter()
+                                .map(|value| match value {
+                                    Value::Regex(regex) => Ok(Pattern::Regex((**regex).clone())),
+                                    Value::Bytes(bytes) => Ok(Pattern::String(
+                                        String::from_utf8_lossy(&bytes).into_owned(),
+                                    )),
+                                    _ => Err("`patterns` must be regular expressions"),
+                                })
+                                .collect::<std::result::Result<Vec<_>, _>>()?),
+                            _ => Err("`patterns` must be array of regular expression literals"),
+                        }?;
+                        Ok(Filter::Pattern(patterns))
+                    }
+                    _ => Err("unknown filter name"),
+                }
+            }
+            Value::Bytes(bytes) => match bytes.as_ref() {
+                b"pattern" => Err("pattern cannot be used without arguments"),
+                b"us_social_security_number" => Ok(Filter::UsSocialSecurityNumber),
+                _ => Err("unknown filter name"),
+            },
+            Value::Regex(regex) => Ok(Filter::Pattern(vec![Pattern::Regex((*regex).clone())])),
+            _ => Err("unknown literal for filter, must be a regex, filter name, or object"),
         }
     }
 }
 
-impl FromStr for Filter {
-    type Err = &'static str;
-
-    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
-        use Filter::*;
-
-        match s {
-            "pattern" => Ok(Pattern),
-            _ => Err("unknown filter"),
+impl Filter {
+    fn redact<'t>(&self, input: Cow<'t, str>, redactor: &Redactor) -> Cow<'t, str> {
+        match &self {
+            Filter::Pattern(patterns) => patterns.iter().fold(input, |input, pattern| {
+                // TODO see if we can avoid cloning here
+                match pattern {
+                    Pattern::Regex(regex) => regex
+                        .replace_all(&input, redactor.pattern())
+                        .into_owned()
```

Comment:
Since the return from this function is `Cow<'t, str>` which is also the return from `replace_all`, I believe the `into_owned().into()` part is unnecessary.

Reply:
👍 thanks for the reminder to poke at this. I'd meant to resolve that TODO earlier.

Unfortunately, I am having a hard time satisfying the borrow checker with removing that `into_owned()`:

```
212 |                         Pattern::Regex(regex) => regex.replace_all(&input, redactor.pattern()),
    |                                                  ^^^^^^^^^^^^^^^^^^------^^^^^^^^^^^^^^^^^^^^^
    |                                                  |                 |
    |                                                  |                 `input` is borrowed here
    |                                                  returns a value referencing data owned by the current function
```

I then thought I'd try changing the input parameter to a `&str` but I eventually ran into the same issue.

In any event, I think it would transparently be converting into an owned string on each `replace_all` call anyway and we need it to be owned at the end to put into a `vrl::Value` so I'm not sure trying to keep it as `Cow` saves us much. I'm curious what you think though.